[Skia] Use the serializer generator for SkString and SkFontStyle::Slant

fujii opened the pull request:
#### 154066861f1aec79e1a444659b7b8dbbd2cd78b2
<pre>
[Skia] Use the serializer generator for SkString and SkFontStyle::Slant
<a href="https://bugs.webkit.org/show_bug.cgi?id=284008">https://bugs.webkit.org/show_bug.cgi?id=284008</a>

Reviewed by Alex Christensen.

Added wrapper classes CoreIPCSkString and CoreIPCSkFontStyleSlant.
Removed the handwritten coders.

* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/skia/CoreIPCSkFontStyleSlant.h: Added.
(WebKit::CoreIPCSkFontStyleSlant::CoreIPCSkFontStyleSlant):
(WebKit::CoreIPCSkFontStyleSlant::create):
(WebKit::CoreIPCSkFontStyleSlant::data const):
* Source/WebKit/Shared/skia/CoreIPCSkFontStyleSlant.serialization.in: Added.
* Source/WebKit/Shared/skia/CoreIPCSkString.h: Added.
(WebKit::CoreIPCSkString::CoreIPCSkString):
(WebKit::CoreIPCSkString::create):
(WebKit::CoreIPCSkString::data const):
* Source/WebKit/Shared/skia/CoreIPCSkString.serialization.in: Added.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp: Removed.

Canonical link: <a href="https://commits.webkit.org/287481@main">https://commits.webkit.org/287481@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/321f8648830b551e5efc9e9b453190103968106f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79860 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/58860 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33258 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/84378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/30850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/67928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7150 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/84378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/30850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82928 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/67928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/72727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/84378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/67928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/29303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/67928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/27351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85808 "Failed to checkout and rebase branch from PR 37397") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/7085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/85808 "Failed to checkout and rebase branch from PR 37397") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/7259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/68567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/85808 "Failed to checkout and rebase branch from PR 37397") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/12849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12344 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7046 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/6900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/10412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/8706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/skia/CoreIPCSkFontStyleSlant.h:
```diff
@@ -0,0 +1,57 @@
+
+    static SkFontStyle::Slant create(uint8_t slant)
+    {
+        return static_cast<SkFontStyle::Slant>(slant);
```
This doesn't prevent out-of-bounds values.  I think it would be better to make an enum class that mirrors Slant's 3 values and use that instead.

fujii replied:
SkFontStyle constructor clamps the input value with `SkTPin` which is using min and max.
https://github.com/WebKit/WebKit/blob/ac3edb570f403e4e1e47bee55f610a99f9d46c38/Source/ThirdParty/skia/include/core/SkFontStyle.h#L53C10-L53C16
https://github.com/WebKit/WebKit/blob/6194732613da51a04d72cb4b25b670c69ab6e663/Source/ThirdParty/skia/include/private/base/SkTPin.h#L20

achristensen07 replied:
Yikes

achristensen07 commented on Source/WebKit/Shared/skia/CoreIPCSkString.h:
```diff
@@ -0,0 +1,57 @@
+    }
+
+private:
+    const SkString& m_string;
```
Storing an unsafe reference isn't ideal.  Are the Skia maintainers amenable to adding a span accessor and constructor to SkString?

fujii replied:
I don't think Skia adopts C++20 `std::span` in their public API.
They have `SkSpan` instead.
https://github.com/WebKit/WebKit/blob/main/Source/ThirdParty/skia/include/private/base/SkSpan_impl.h
Should I ask the upstream Skia to add a `SkString` constructor which takes `SkSpan<const char>`?

fujii replied:
I think raw reference is inevitable for lvalue-reference-taking encoders of external types for performance.

fujii committed:
[Skia] Use the serializer generator for SkString and SkFontStyle::Slant
https://bugs.webkit.org/show_bug.cgi?id=284008

Reviewed by Alex Christensen.

Added wrapper classes CoreIPCSkString and CoreIPCSkFontStyleSlant.
Removed the handwritten coders.

* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/skia/CoreIPCSkFontStyleSlant.h: Added.
(WebKit::CoreIPCSkFontStyleSlant::CoreIPCSkFontStyleSlant):
(WebKit::CoreIPCSkFontStyleSlant::create):
(WebKit::CoreIPCSkFontStyleSlant::data const):
* Source/WebKit/Shared/skia/CoreIPCSkFontStyleSlant.serialization.in: Added.
* Source/WebKit/Shared/skia/CoreIPCSkString.h: Added.
(WebKit::CoreIPCSkString::CoreIPCSkString):
(WebKit::CoreIPCSkString::create):
(WebKit::CoreIPCSkString::data const):
* Source/WebKit/Shared/skia/CoreIPCSkString.serialization.in: Added.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp: Removed.

Canonical link: https://commits.webkit.org/287481@main
